<a href="https://colab.research.google.com/github/hanghae-plus-AI/AI-1-soyoungcareer/blob/main/week5/Chapter3_2_%EA%B8%B0%EB%B3%B8%EA%B3%BC%EC%A0%9C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 실습: Zero-shot Classification

이번 실습에서는 open LLM을 가지고 zero-shot classification을 해봅니다. 먼저 필요한 library들을 설치합시다.

In [ ]:
!pip -q install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.6/471.6 kB 23.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 11.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 13.0 MB/s eta 0:00:00


그 다음 Gemma-2B를 사용하기 위해 다음과 같은 작업을 진행합니다:
1. huggingface.co 계정 만들고 로그인하기
2. https://www.kaggle.com/models/google/gemma/license/consent 에서 Gemma license 동의하기
3. 홈 화면으로 돌아와, `Profile > Settings > Access Tokens` 메뉴로 들어와 "Write" type의 token 생성하기
4. 생성한 토큰을 아래 "HF TOKEN"에 불여넣고 셀을 실행하기.

In [ ]:
from huggingface_hub import login
from google.colab import userdata


login(userdata.get('hf_token'))

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /root/.cache/huggingface/token
Login successful


정상적으로 token을 생성하고 Gemma license에 동의했다면 아래 코드로 tokenizer와 Gemma-2B 모델을 불러올 수 있습니다.

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/33.6k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/636 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/627 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/13.5k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/67.1M [00:00<?, ?B/s]

`config.hidden_act` is ignored, you should use `config.hidden_activation` instead.
Gemma's activation function will be set to `gelu_pytorch_tanh`. Please, use
`config.hidden_activation` if you want to override this behaviour.
See https://github.com/huggingface/transformers/pull/29402 for more details.


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

이번에는 Gemma-2B를 가지고 간단한 text 생성을 해봅시다.
"What is your name?" 이라는 text를 넣었을 때 어떤 text가 생성되는지 살펴봅시다.

In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
outputs = model.generate(**input_ids)
print(tokenizer.decode(outputs[0]))

/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1258: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


<bos>What is your name?

What is your age?

What is your gender?

What


2B의 작은 LLM이라 질좋은 답변이 나오지 않는 것을 알 수 있습니다.
이번에는 입력으로 넣어준 token들의 logit을 계산해봅시다.

In [ ]:
tokens = input_ids['input_ids']
print(tokens)

logits = model(**input_ids).logits
for i in range(tokens.shape[-1]):
    token = tokens[0, i].item()
    print(logits[0, i, token])

tensor([[     2,   1841,    603,    861,   1503, 235336]], device='cuda:0')
tensor(-18.2747, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(-33.2665, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(-23.9536, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(-27.7627, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(-19.6064, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(-21.0372, device='cuda:0', grad_fn=<SelectBackward0>)


위와 같이 모델 출력의 `.logits`을 통해 token들의 logit을 알 수 있습니다.
Logit은 높을 수록 token이 나올 확률이 높다는 뜻입니다.

이번에는 logit 계산을 통해 zero-shot classification을 구현해보도록 하겠습니다.

In [ ]:
import torch

def zero_shot_classification(text, task_description, labels):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to("cuda")  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.
    probs = []
    for label in labels:  # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("cuda")
        n_label_tokens = label_ids['input_ids'].shape[-1] - 1  # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            'input_ids': torch.concatenate([text_ids['input_ids'], label_ids['input_ids'][:, 1:]], axis=-1),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            'attention_mask': torch.concatenate([text_ids['attention_mask'], label_ids['attention_mask'][:, 1:]], axis=-1)
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids['input_ids'].shape[-1]
        for i in range(n_label_tokens, 0, -1):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids['input_ids'][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.cuda.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

아래는 실제로 zero-shot classification을 해본 결과입니다.

In [ ]:
probs = zero_shot_classification("I am happy!", "Is the sentence positive or negative?: ", ["positive", "negative"])
print(probs)

[-4.51517391204834, -9.590051651000977]


보시다시피 우리는 Gemma를 별도로 학습하지 않았음에도 불구하고 주어진 문장이 긍정적이라는 것을 정확하게 예측하고 있습니다.

다음은 영화 리뷰 감정 분석 task에 적용해봅시다.
먼저 data를 불러옵니다.

In [ ]:
from datasets import load_dataset
from transformers import pipeline

# 데이터셋 변경
ag_news = load_dataset("fancyzhx/ag_news")

# 레이블 확인
label_names = ag_news['train'].features['label'].names
print(label_names)

def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=512, truncation=True)

tokenized_ag_news = ag_news.map(preprocess_function, batched=True)

['World', 'Sports', 'Business', 'Sci/Tech']


Map:   0%|          | 0/120000 [00:00<?, ? examples/s]

Map:   0%|          | 0/7600 [00:00<?, ? examples/s]

그리고 `test` data에서 50개의 영화 리뷰에 대해 예측하는 코드는 다음과 같습니다.

In [ ]:
import numpy as np
from tqdm import tqdm


n_corrects = 0
n_total = 50
for i in tqdm(range(n_total)):
    text = tokenized_ag_news['test'][i]['text']
    label = tokenized_ag_news['test'][i]['label']
    probs = zero_shot_classification(
        text,
        "A news article is given. Decide which category it belongs to : ",
        labels=["Answer: World.", "Answer: Sports.", "Answer: Business.", "Answer: Sci/Tech."]
    )

    print(probs)

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1

print(n_corrects)

accuracy = n_corrects/ n_total
print(f"Accuracy : {accuracy}")

  2%|▏         | 1/50 [00:00<00:13,  3.53it/s]

[-55.41572189331055, -53.861202239990234, -51.574435234069824, -58.57706832885742]


  4%|▍         | 2/50 [00:00<00:18,  2.63it/s]

[-48.31380844116211, -45.561734199523926, -46.98854446411133, -46.821237325668335]


  6%|▌         | 3/50 [00:01<00:18,  2.54it/s]

[-52.890503883361816, -52.49685764312744, -46.83700942993164, -49.529683113098145]


  8%|▊         | 4/50 [00:01<00:18,  2.42it/s]

[-52.43497562408447, -54.77520751953125, -54.728912353515625, -55.13695949316025]


 10%|█         | 5/50 [00:01<00:18,  2.44it/s]

[-47.32776641845703, -51.015602111816406, -48.98029708862305, -50.529332995414734]


 12%|█▏        | 6/50 [00:02<00:22,  1.97it/s]

[-55.2263879776001, -57.134446144104004, -53.77957248687744, -56.42442846298218]


 14%|█▍        | 7/50 [00:03<00:26,  1.63it/s]

[-55.18997097015381, -56.99260139465332, -55.72818660736084, -57.02354574203491]


 16%|█▌        | 8/50 [00:04<00:27,  1.55it/s]

[-54.37621307373047, -61.06999588012695, -61.21863555908203, -59.82126522064209]


 18%|█▊        | 9/50 [00:04<00:21,  1.89it/s]

[-54.06928253173828, -55.862382888793945, -51.8411340713501, -51.81559085845947]


 20%|██        | 10/50 [00:04<00:19,  2.03it/s]

[-57.25523662567139, -57.07066345214844, -54.48747444152832, -59.63089174032211]


 22%|██▏       | 11/50 [00:05<00:19,  2.03it/s]

[-49.542073249816895, -47.34271812438965, -45.94670486450195, -46.289421796798706]


 24%|██▍       | 12/50 [00:05<00:18,  2.10it/s]

[-50.93826675415039, -46.018961906433105, -44.915876388549805, -46.02516722679138]


 26%|██▌       | 13/50 [00:06<00:16,  2.19it/s]

[-48.79990863800049, -47.80779552459717, -44.182634353637695, -46.43563365936279]


 28%|██▊       | 14/50 [00:06<00:14,  2.47it/s]

[-50.13406848907471, -52.11188793182373, -50.76607608795166, -52.193183958530426]


 30%|███       | 15/50 [00:06<00:12,  2.73it/s]

[-52.53021812438965, -52.14429759979248, -52.766693115234375, -55.72311681509018]


 32%|███▏      | 16/50 [00:07<00:11,  2.95it/s]

[-50.91747856140137, -49.660945892333984, -50.58914375305176, -47.96334421634674]


 34%|███▍      | 17/50 [00:07<00:11,  2.75it/s]

[-55.70283603668213, -59.301950454711914, -60.67468547821045, -58.44702725112438]


 36%|███▌      | 18/50 [00:07<00:12,  2.62it/s]

[-59.18484878540039, -62.17163276672363, -62.64358711242676, -63.344944570213556]


 38%|███▊      | 19/50 [00:08<00:12,  2.56it/s]

[-56.31660747528076, -61.37694835662842, -61.085591316223145, -59.41330623626709]


 40%|████      | 20/50 [00:08<00:10,  2.79it/s]

[-62.12232208251953, -63.86345958709717, -58.37561511993408, -65.61938810348511]


 42%|████▏     | 21/50 [00:08<00:09,  2.96it/s]

[-55.50384521484375, -55.66769599914551, -51.87911510467529, -53.87474536895752]


 44%|████▍     | 22/50 [00:09<00:08,  3.13it/s]

[-56.111074447631836, -53.13184642791748, -52.25406360626221, -51.169589042663574]


 46%|████▌     | 23/50 [00:09<00:08,  3.29it/s]

[-57.57218265533447, -59.00341987609863, -55.76551532745361, -53.71175980567932]


 48%|████▊     | 24/50 [00:09<00:09,  2.85it/s]

[-54.23295593261719, -54.56105422973633, -50.950151443481445, -53.766101121902466]


 50%|█████     | 25/50 [00:10<00:09,  2.70it/s]

[-57.97323036193848, -60.2360782623291, -57.84079360961914, -62.370786905288696]


 52%|█████▏    | 26/50 [00:10<00:08,  2.91it/s]

[-56.47314262390137, -57.00347328186035, -56.714919090270996, -60.99012351036072]


 54%|█████▍    | 27/50 [00:10<00:07,  3.02it/s]

[-43.58660411834717, -42.88572311401367, -47.39517307281494, -48.35628664493561]


 56%|█████▌    | 28/50 [00:11<00:08,  2.61it/s]

[-61.91734504699707, -56.061158180236816, -68.8366641998291, -65.39617371559143]


 58%|█████▊    | 29/50 [00:11<00:08,  2.47it/s]

[-58.37406349182129, -52.980475425720215, -58.571340560913086, -61.59087872505188]


 60%|██████    | 30/50 [00:12<00:08,  2.37it/s]

[-58.51954174041748, -54.23537063598633, -60.654311180114746, -60.41306662559509]


 62%|██████▏   | 31/50 [00:12<00:07,  2.38it/s]

[-53.648701667785645, -50.29997730255127, -55.25108814239502, -56.268678188323975]


 64%|██████▍   | 32/50 [00:13<00:06,  2.58it/s]

[-46.97680187225342, -43.24625873565674, -48.12655067443848, -51.56327927112579]


 66%|██████▌   | 33/50 [00:13<00:06,  2.53it/s]

[-49.85377883911133, -50.26827049255371, -50.33306407928467, -57.364811420440674]


 68%|██████▊   | 34/50 [00:13<00:06,  2.42it/s]

[-51.73671627044678, -49.91376495361328, -49.46299934387207, -56.91873836517334]


 70%|███████   | 35/50 [00:14<00:06,  2.36it/s]

[-54.642924308776855, -60.168556213378906, -59.63632678985596, -59.11323839426041]


 72%|███████▏  | 36/50 [00:14<00:06,  2.30it/s]

[-56.37182426452637, -55.88696479797363, -60.2447452545166, -63.00159430503845]


 74%|███████▍  | 37/50 [00:15<00:05,  2.57it/s]

[-49.23569297790527, -52.53436470031738, -50.3418493270874, -56.30083155632019]


 76%|███████▌  | 38/50 [00:15<00:04,  2.51it/s]

[-50.63406753540039, -49.416470527648926, -46.287962913513184, -49.51324498653412]


 78%|███████▊  | 39/50 [00:15<00:04,  2.48it/s]

[-46.976853370666504, -46.174795150756836, -48.21469020843506, -46.25540333986282]


 80%|████████  | 40/50 [00:16<00:04,  2.44it/s]

[-53.811964988708496, -56.959805488586426, -58.06668567657471, -60.80125904083252]


 82%|████████▏ | 41/50 [00:16<00:03,  2.42it/s]

[-44.23209571838379, -40.82970428466797, -45.196340560913086, -48.08729028701782]


 84%|████████▍ | 42/50 [00:17<00:03,  2.37it/s]

[-50.70153999328613, -52.57309055328369, -48.112449645996094, -53.74213123321533]


 86%|████████▌ | 43/50 [00:17<00:02,  2.38it/s]

[-49.715322494506836, -50.09645938873291, -51.52739906311035, -53.71380579471588]


 88%|████████▊ | 44/50 [00:18<00:02,  2.33it/s]

[-49.80439281463623, -49.41424560546875, -45.15952014923096, -47.59712862968445]


 90%|█████████ | 45/50 [00:18<00:02,  2.30it/s]

[-45.77958011627197, -42.879226207733154, -48.62729263305664, -50.34847116470337]


 92%|█████████▏| 46/50 [00:18<00:01,  2.53it/s]

[-47.67562294006348, -45.56596851348877, -49.79941749572754, -55.995824217796326]


 94%|█████████▍| 47/50 [00:19<00:01,  2.48it/s]

[-40.45131969451904, -39.29545593261719, -43.81350135803223, -44.676753759384155]


 96%|█████████▌| 48/50 [00:19<00:00,  2.40it/s]

[-46.97861194610596, -47.970234870910645, -42.117473125457764, -44.023744344711304]


 98%|█████████▊| 49/50 [00:20<00:00,  2.39it/s]

[-44.18174934387207, -43.57655906677246, -42.99946212768555, -43.929457664489746]


100%|██████████| 50/50 [00:20<00:00,  2.43it/s]

[-52.97128105163574, -49.6692008972168, -52.66174507141113, -54.153380393981934]
21
Accuracy : 0.42


보시다시피 정확도 88%로, 매우 높은 성능을 보이는 것을 알 수 있습니다.